File: src/main/java/net/maple/packets/ConversationPackets.java
package net.maple.packets;

import net.maple.SendOpcode;
import scripting.npc.ConversationType;
import util.packet.Packet;
import util.packet.PacketWriter;

public class ConversationPackets {

    private static PacketWriter getMessagePacket(int npc, ConversationType type, int speaker, String text) {
        PacketWriter pw = new PacketWriter(16);

        pw.writeHeader(SendOpcode.SCRIPT_MESSAGE);
        pw.write(0); // SpeakerTypeID
        pw.writeInt(npc);
        pw.write(type.getValue());
        pw.write(speaker);
        pw.writeMapleString(text);

        return pw;
    }

    private static Packet getSayMessagePacket(int npc, int speaker, String text, boolean prev, boolean next) {
        PacketWriter pw = getMessagePacket(npc, ConversationType.SAY, speaker, text);

        pw.writeBool(prev);
        pw.writeBool(next);

        return pw.createPacket();
    }

    public static Packet getOkMessagePacket(int npc, int speaker, String text) {
        return getSayMessagePacket(npc, speaker, text, false, false);
    }

    public static Packet getPrevMessagePacket(int npc, int speaker, String text) {
        return getSayMessagePacket(npc, speaker, text, true, false);
    }

    public static Packet getNextMessagePacket(int npc, int speaker, String text) {
        return getSayMessagePacket(npc, speaker, text, false, true);
    }

    public static Packet getNextPrevMessagePacket(int npc, int speaker, String text) {
        return getSayMessagePacket(npc, speaker, text, true, true);
    }

    public static Packet getYesNoMessagePacket(int npc, int speaker, String text) {
        return getMessagePacket(npc, ConversationType.ASK_YES_NO, speaker, text).createPacket();
    }

    public static Packet getTextMessagePacket(int npc, int speaker, String text, String def, int min, int max) {
        PacketWriter pw = getMessagePacket(npc, ConversationType.ASK_TEXT, speaker, text);
        pw.writeMapleString(def);
        pw.writeShort(min);
        pw.writeShort(max);

        return pw.createPacket();
    }

    public static Packet getNumberMessagePacket(int npc, int speaker, String text, int def, int min, int max) {
        PacketWriter pw = getMessagePacket(npc, ConversationType.ASK_NUMBER, speaker, text);
        pw.writeInt(def);
        pw.writeInt(min);
        pw.writeInt(max);

        return pw.createPacket();
    }

    public static Packet getSimpleMessagePacket(int npc, int speaker, String text) {
        return getMessagePacket(npc, ConversationType.ASK_MENU, speaker, text).createPacket();
    }

    public static Packet getAcceptMessagePacket(int npc, int speaker, String text) {
        return getMessagePacket(npc, ConversationType.ASK_ACCEPT, speaker, text).createPacket();
    }

    public static Packet getBoxTextMessagePacket(int npc, int speaker, String def, int cols, int rows) {
        PacketWriter pw = getMessagePacket(npc, ConversationType.ASK_BOX_TEXT, speaker, "");
        pw.writeMapleString(def);
        pw.writeShort(cols);
        pw.writeShort(rows);

        return pw.createPacket();
    }

    /**
     * Very weird one, dimensional portal in gms I think
     */
    public static Packet getSlideMenuMessagePacket(int npc, int speaker, String text, int type, int selected) {
        PacketWriter pw = new PacketWriter(16);

        pw.writeHeader(SendOpcode.SCRIPT_MESSAGE);
        pw.write(0); // SpeakerTypeID
        pw.writeInt(npc);
        pw.write(ConversationType.ASK_SLIDE_MENU.getValue());
        pw.write(speaker);

        pw.writeInt(type);
        pw.writeInt(selected);
        pw.writeMapleString(text);

        return pw.createPacket();
    }
}

File: src/main/java/net/maple/packets/ConversationPackets.kt
package net.maple.packets

import net.maple.SendOpcode
import scripting.dialog.ConversationType
import scripting.dialog.SpeakerType
import util.packet.Packet
import util.packet.PacketWriter

object ConversationPackets {

    private fun getMessagePacket(npc: Int, type: ConversationType, speaker: Int, text: String, replaceNpc: Int?): PacketWriter {
        val pw = PacketWriter(16)

        pw.writeHeader(SendOpcode.SCRIPT_MESSAGE)
        pw.write(0) // SpeakerTypeID
        pw.writeInt(npc)
        pw.write(type.value)
        pw.write(speaker)

        if (speaker and SpeakerType.NpcReplacedByNpc == SpeakerType.NpcReplacedByNpc) {
            pw.writeInt(replaceNpc ?: npc)
        }

        pw.writeMapleString(text)

        return pw
    }

    private fun getSayMessagePacket(npc: Int, speaker: Int, text: String, prev: Boolean, next: Boolean, replaceNpc: Int?): Packet {
        val pw = getMessagePacket(npc, ConversationType.SAY, speaker, text, replaceNpc)

        pw.writeBool(prev)
        pw.writeBool(next)

        return pw.createPacket()
    }

    fun getOkMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getSayMessagePacket(npc, speaker, text, prev = false, next = false, replaceNpc)
    }

    fun getPrevMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getSayMessagePacket(npc, speaker, text, prev = true, next = false, replaceNpc)
    }

    fun getNextMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getSayMessagePacket(npc, speaker, text, prev = false, next = true, replaceNpc)
    }

    fun getNextPrevMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getSayMessagePacket(npc, speaker, text, prev = true, next = true, replaceNpc)
    }

    fun getYesNoMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getMessagePacket(npc, ConversationType.ASK_YES_NO, speaker, text, replaceNpc).createPacket()
    }

    fun getTextMessagePacket(npc: Int, speaker: Int, text: String, def: String, min: Int, max: Int, replaceNpc: Int? = null): Packet {
        val pw = getMessagePacket(npc, ConversationType.ASK_TEXT, speaker, text, replaceNpc)

        pw.writeMapleString(def)
        pw.writeShort(min)
        pw.writeShort(max)

        return pw.createPacket()
    }

    fun getNumberMessagePacket(npc: Int, speaker: Int, text: String, def: Int, min: Int, max: Int, replaceNpc: Int? = null): Packet {
        val pw = getMessagePacket(npc, ConversationType.ASK_NUMBER, speaker, text, replaceNpc)

        pw.writeInt(def)
        pw.writeInt(min)
        pw.writeInt(max)

        return pw.createPacket()
    }

    fun getSimpleMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getMessagePacket(npc, ConversationType.ASK_MENU, speaker, text, replaceNpc).createPacket()
    }

    fun getAcceptMessagePacket(npc: Int, speaker: Int, text: String, replaceNpc: Int? = null): Packet {
        return getMessagePacket(npc, ConversationType.ASK_ACCEPT, speaker, text, replaceNpc).createPacket()
    }

    fun getBoxTextMessagePacket(npc: Int, speaker: Int, def: String, cols: Int, rows: Int, replaceNpc: Int? = null): Packet {
        val pw = getMessagePacket(npc, ConversationType.ASK_BOX_TEXT, speaker, "", replaceNpc)

        pw.writeMapleString(def)
        pw.writeShort(cols)
        pw.writeShort(rows)

        return pw.createPacket()
    }

    /**
     * Very weird one, dimensional portal in gms I think
     */
    fun getSlideMenuMessagePacket(npc: Int, speaker: Int, text: String, type: Int, selected: Int): Packet {
        val pw = PacketWriter(16)

        pw.writeHeader(SendOpcode.SCRIPT_MESSAGE)
        pw.write(0) // SpeakerTypeID
        pw.writeInt(npc)
        pw.write(ConversationType.ASK_SLIDE_MENU.value)
        pw.write(speaker)
        pw.writeInt(type)
        pw.writeInt(selected)
        pw.writeMapleString(text)

        return pw.createPacket()
    }
}